flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,235 @@
+        return jvmArgs;
+    }
+
+    public void includeLibraries(FileCollection fileCollection) {
+        into(Flask.Constants.LIBRARIES_FOLDER, (copySpec) -> copySpec.from(fileCollection));
+    }
+
+    private static class JavaAgent {
+        File jar;
+        String args;
+    }
+
+    private final List<JavaAgent> javaAgents;
+
+    @InputFiles
+    FileCollection getAgentJars() {
+        File[] jars = javaAgents.stream().map(it -> it.jar).toArray(File[]::new);
+        return getProject().files((Object[]) jars);
+    }
+
+    @Input
+    List<String> getAgentArgs() {
+        return javaAgents.stream().map(it -> it.args).collect(Collectors.toList());
+    }
+
+    public void javaAgent(Action<JavaAgent> action) {
+        JavaAgent agent = new JavaAgent();
+        action.execute(agent);
+        if(agent.jar == null) throw new GradleException("No jar file specified for Java agent");
+        javaAgents.add(agent);
+    }
+
+    public FlaskJarTask() {
+        BasePluginConvention basePluginConvention = getProject().getConvention().getPlugin(BasePluginConvention.class);
+        getDestinationDirectory().set(basePluginConvention.getLibsDir());
+        getArchiveBaseName().convention(getProject().getName());
+        getArchiveExtension().convention("jar");
+        ObjectFactory objects = getProject().getObjects();
+        launcherClassName = objects.property(String.class).convention(Flask.Constants.DEFAULT_LAUNCHER_NAME);
+        mainClassName = objects.property(String.class);
+        jvmArgs = objects.listProperty(String.class).convention(new ArrayList<>());
+        javaAgents = new ArrayList<>();
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
+    }
+
+    @Input
+    @SneakyThrows
+    public String getLauncherArchiveHash() {
+        MessageDigest md5 = MessageDigest.getInstance("MD5");
+        try(InputStream stream = new DigestInputStream(LauncherResource.instance.read(), md5)) {
+            byte[] buffer = new byte[0x10000];
+            while(stream.read(buffer) >= 0) {}
+        }
+        return String.format("%032x", new BigInteger(1, md5.digest()));
+    }
+
+    @RequiredArgsConstructor
+    private static class StreamAction implements CopyActionProcessingStreamAction {
+
+        private final ZipOutputStream zoos;
+        private final Manifest manifest;
+        private final MessageDigest md;
+        private final byte[] buffer;
+
+        private final List<FileCopyDetailsInternal> fileCopyDetailsInternals = new ArrayList<>();
+
+        @SneakyThrows
+        public void write() {
+            ZipEntry zipEntry = new ZipEntry(Flask.Constants.METADATA_FOLDER + '/');
+            zoos.putNextEntry(zipEntry);
+            zipEntry = new ZipEntry(JarFile.MANIFEST_NAME);
+            zipEntry.setMethod(ZipEntry.DEFLATED);
+            zoos.putNextEntry(zipEntry);
+            manifest.write(zoos);
+            zoos.closeEntry();
```

Comment:
FYI calling `putNextEntry()` will automatically invoke `closeEntry()` for the previous one,